_source/_posts/2021-12-06-spring-data-elasticsearch.md

```diff
@@ -0,0 +1,349 @@
+
+JHipster provides the Elasticsearch option to add search capabilities on top of your database. The integration is based on Spring Data Elasticsearch repositories, so let's generate a reactive Blog application to explore what the generator provides. The Blog application built in this tutorial is based on the JHipster sample [reactive-ms.jdl](https://github.com/jhipster/jdl-samples/blob/main/reactive-ms.jdl), but adding Maven, MongoDB (Elasticsearch integration only works with SQL databases and MongoDB), Bootstrap pagination, Okta authentication, and Kibana for index mapping inspection.
+
+Start by creating a `spring-data-elasticsearch` folder for the project, get the application JDL from the [Github](https://github.com/indiepopart/spring-data-elasticsearch.git) repository, and generate the application with JHipster:
+
+```shell
+mkdir spring-data-elasticsearch
+cd spring-data-elasticsearch
+https -d https://raw.githubusercontent.com/indiepopart/spring-data-elasticsearch/5380a940a0364176a5d56349d3471f60a0834ed7/blog-reactive-ms.jdl
+jhipster jdl blog-reactive-ms.jdl
+```
+
+{% include setup/cli.md type="jhipster" %}
+
+Follow the process above for creating 2 applications in Okta, the gateway application, and the blog application.
+
+```shell
+cd gateway
+okta apps create jhipster
+```
+
+Make sure to set up http://localhost:8081/login/oauth2/code/oidc with port 8081 as Redirect URI for blog microservice.
+
+```shell
+cd blog
+okta apps create jhipster
+```
+
+Find the `blog` credentials in the files `blog/.okta.env` and  `gateway/.okta.env`.
+
+The JHipster Registry is also a Spring Cloud Config server, and by default, it is configured with the profiles `dev` and `native`, which means the configuration will be provided from the location `docker-compose/central-server/config`. Let's add some files so it can provide the OAuth2 configuration to the microservices.
+Add the file `docker-compose/central-server-config/blog-prod.yml` with the following content:
+
+```yml
+spring:
+  security:
+    oauth2:
+      client:
+        provider:
+          oidc:
+            issuer-uri: https://{yourOktaDomain}/oauth2/default
+        registration:
+          oidc:
+            client-id: {blogClientId}
+            client-secret: {blogClientSecret}
+```
+
+Add also the file `docker-compose/central-server-config/gateway-prod.yml` with the following content:
+
+```yml
+spring:
+  security:
+    oauth2:
+      client:
+        provider:
+          oidc:
+            issuer-uri: https://{yourOktaDomain}/oauth2/default
+        registration:
+          oidc:
+            client-id: {gatewayClientId}
+            client-secret: {gatewayClientSecret}
+```
+
+For the experimentation in this tutorial, configure the Kibana interface for Elasticsearch, which will allow visualizations of the Elasticsearch data.
+Edit `docker-compose/docker-compose.yml` and add the Kibana service like this:
+
+```yml
+services:
```

Comment:
Add `...` on its own line to help indicate this is a top-level key.

Reply:
Done